votepy/generic_ilp.py

```diff
@@ -0,0 +1,141 @@
+from abc import ABC,abstractmethod
+from enum import Enum
+from typing import Optional
+
+class Solver(ABC):
+    class Sense(Enum):
+        MAX = 1
+        MIN = 2
+    @abstractmethod
+    def addVariable(self, name: str, vartype: str, obj: Optional[float], lb: Optional[float], ub: Optional[float]):
+        pass
+
+    @abstractmethod
+    def addConstraint(self, name: str, variables: list[object], coeffs: list[float], rhs: float, sense: str):
+        pass
+
+    @abstractmethod
+    def solve(self):
+        pass
+
+    @abstractmethod
+    def getValues(self):
+        pass
+
+
+class CPLEX(Solver):
+    def __init__(self, sense: Solver.Sense = Solver.Sense.MAX) -> None:
+        super().__init__()
+        try:
+            import cplex
+        except ImportError as err:
+            print("[Error]: Failed to import {}".format(err.args[0]))
+            exit(1)
+
+        self.colnames = []
+        self.obj = []
+        self.lb = []
+        self.ub = []
+        self.ctypes = []
+
+        self.rhs = []
+        self.rownames = []
+        self.senses = []
+        self.rows = []
+
+        self.cplex = cplex
+        self.model = cplex.Cplex()
+        self.model.set_log_stream(None)
+        self.model.set_results_stream(None)
+
+
+        if sense == Solver.Sense.MAX:
+            self.model.objective.set_sense(self.model.objective.sense.maximize)
+        elif sense == Solver.Sense.MIN:
+            self.model.objective.set_sense(self.model.objective.sense.minimize)
+
+
+    def addVariable(self, name: str, vartype: str, obj: float = 0.0, lb: Optional[float] = None, ub: Optional[float] = None):
+        self.colnames.append(name)
+        self.obj.append(obj)
+        if lb is not None:
+            self.lb.append(lb)
+        if ub is not None:
+            self.ub.append(ub)
+        self.ctypes.append(vartype)
+        return name
+
+    def addConstraint(self, name: str, variables: list[object], coeffs: list[float], rhs: float, sense: str):
+        self.rownames.append(name)
+        self.rows.append([variables, coeffs])
+        self.rhs.append(rhs)
+        self.senses.append(sense)
+
+    def solve(self):
+        if len(self.lb) != len(self.ctypes) or len(self.ub) != len(self.ctypes):
+            if len(self.lb) > 0 or len(self.ub) > 0:
+                print("Warning: You've given bounds for some variables but not for all - please make sure you either set bounds for every variable or for none.\nUsing default bounds...")
```

Comment:
Maybe we should use the `warnings` module from the standard librabry? https://docs.python.org/3/library/warnings.html

Reply:
Looks good, let's do that